src/openrct2/drawing/Drawing.Sprite.cpp

```diff
@@ -358,59 +358,43 @@ bool gfx_load_csg()
     }
 }
 
-const uint8_t* FASTCALL gfx_draw_sprite_get_palette(ImageId imageId)
+static std::optional<PaletteMap> FASTCALL gfx_draw_sprite_get_palette(ImageId imageId)
 {
     if (!imageId.HasSecondary())
     {
-        uint8_t palette_ref = imageId.GetRemap();
+        uint8_t paletteId = imageId.GetRemap();
         if (!imageId.IsBlended())
         {
-            palette_ref &= 0x7F;
-        }
-
-        uint16_t palette_offset = palette_to_g1_offset[palette_ref];
-        auto g1 = gfx_get_g1_element(palette_offset);
-        if (g1 == nullptr)
-        {
-            return nullptr;
-        }
-        else
-        {
-            return g1->offset;
+            paletteId &= 0x7F;
         }
+        return GetPaletteMapForColour(paletteId);
     }
     else
     {
-        uint8_t* palette_pointer = gPeepPalette;
-
-        uint32_t primary_offset = palette_to_g1_offset[imageId.GetPrimary()];
-        uint32_t secondary_offset = palette_to_g1_offset[imageId.GetSecondary()];
-
+        auto paletteMap = PaletteMap(gPeepPalette);
         if (imageId.HasTertiary())
         {
-            palette_pointer = gOtherPalette;
-#if defined(DEBUG_LEVEL_2) && DEBUG_LEVEL_2
-            assert(tertiary_colour < PALETTE_TO_G1_OFFSET_COUNT);
-#endif // DEBUG_LEVEL_2
-            uint32_t tertiary_offset = palette_to_g1_offset[imageId.GetTertiary()];
-            auto tertiary_palette = gfx_get_g1_element(tertiary_offset);
-            if (tertiary_palette != nullptr)
+            paletteMap = PaletteMap(gOtherPalette);
+            auto tertiaryPaletteMap = GetPaletteMapForColour(imageId.GetTertiary());
+            if (tertiaryPaletteMap)
             {
-                std::memcpy(palette_pointer + 0x2E, &tertiary_palette->offset[0xF3], 12);
+                paletteMap.Copy(0x2E, *tertiaryPaletteMap, 0xF3, 12);
             }
         }
-        auto primary_palette = gfx_get_g1_element(primary_offset);
-        if (primary_palette != nullptr)
+
+        auto primaryPaletteMap = GetPaletteMapForColour(imageId.GetPrimary());
+        if (primaryPaletteMap)
         {
-            std::memcpy(palette_pointer + 0xF3, &primary_palette->offset[0xF3], 12);
+            paletteMap.Copy(0xF3, *primaryPaletteMap, 0xF3, 12);
```

Comment:
The starts of the remap colours should be put into constants. I don't know if there is a constant for the number of shades per colour - it should either be used, or one should be created.